packages/toolkit/src/createAsyncThunk.ts

```diff
@@ -177,7 +177,7 @@ type GetSerializedErrorType<ThunkApiConfig> = ThunkApiConfig extends {
   ? GetSerializedErrorType
   : SerializedError
 
-type MaybePromise<T> = T | Promise<T>
+type MaybePromise<T> = T | Promise<T> | (T extends any ? Promise<T> : never)
```

Comment:
❓ 

For my own learning, do you know *why* it works with this change, but not otherwise? I'm guessing it's something to do with `IsUnknown`, but my brain is turning to jelly trying to understand this

Reply:
https://artsy.github.io/blog/2018/11/21/conditional-types-in-typescript/ - look for this paragraph:

> All we did is wrap the meat of our logic in a flimsy tortilla of inevitability, since the outer condition <code>A extends any</code> will, of course, always be true

In this case, assume

```ts
type Union = 'A' | 'B'

type X = Promise<Union> // will be Promise<'A' | 'B'>
type Y = Union extends any ? Promise<Union> : never // will be Promise<'A'> | Promise<'B'>

```

Reply:
aha, so a case of this then: https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types

I'll probably have to encounter this a couple more times before it sinks in to me that it occurs with `boolean` too (which I guess is 'just' a union of `true | false`.

Thanks for the explanation & link, I'll go over that article in full to brush up my knowledge

Reply:
Yeah, that is happens with `boolean` was completely new to me as well.

Reply:
I just have one question about this now: Isn't `T extends any ? Promise<T> : never` essentially the only type now? `T` will always extend `any` so this will always be `Promise<T>` and not `T`, right? I'm probably missing something.

Reply:
@bfricka it would have to look like this to fit the behaviour you describe:

```ts
type MaybePromise<T> = T extends any ? Promise<T> : T
```

As it is, it's still a union of three possibilities. It might make more sense if you format it a bit differently:
```ts
type MaybePromise<T> = 
  | T
  | Promise<T>
  | (T extends any ? Promise<T> : never)
```

So the fact that `T extends any` is only relevant for the third scenario. The other unions of `T` and `Promise<T>` don't care whether `T extends any` is true or not